File: src/09.tree/19.515.ts
import { TreeNode } from "./01.100";

// BFS
export function largestValues1(root: TreeNode | null): number[] {
    const ans: number[] = [];

    if (!root) {
        return ans;
    }

    const queue: TreeNode[] = [root];

    while (queue.length) {
        let size = queue.length;
        let max = -Infinity;
        while (size > 0) {
            const node = queue.shift();
            max = Math.max(max, node.val);

            node.left && queue.push(node.left);
            node.right && queue.push(node.right);
            size--;
        }
        ans.push(max);
    }

    return ans;
}

// DFS
export function largestValues(root: TreeNode | null): number[] {
    const ans: number[] = [];
    const map = new Map<number, number>();
    const dfs = (node: TreeNode | null, depth: number) => {
        if (!node) {
            return;
        }

        if (!map.has(depth)) {
            map.set(depth, node.val);
        }

        map.set(depth, Math.max(map.get(depth), node.val));

        dfs(node.left, depth + 1);
        dfs(node.right, depth + 1);
    };

    dfs(root, 0);

    map.forEach((val, key) => {
        ans[key] = val;
    });

    return ans;
}

File: src/09.tree/19.515.go
package tree

// BFS
func LargestValues1(root *TreeNode) []int {
	ans := []int{}
	if root == nil {
		return ans
	}

	queue := []*TreeNode{root}

	for len(queue) > 0 {
		max := queue[0].Val
		for i := len(queue); i > 0; i-- {
			node := queue[0]
			queue = queue[1:]

			if node.Val > max {
				max = node.Val
			}

			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		ans = append(ans, max)
	}

	return ans
}

// DFS
func LargestValues(root *TreeNode) []int {
	var dfs func(node *TreeNode, depth int)
	m := map[int]int{}

	dfs = func(node *TreeNode, depth int) {
		if node == nil {
			return
		}

		if _, ok := m[depth]; !ok {
			m[depth] = node.Val
		}

		if node.Val > m[depth] {
			m[depth] = node.Val
		}

		dfs(node.Left, depth+1)
		dfs(node.Right, depth+1)
	}

	dfs(root, 0)

	ans := make([]int, len(m))

	for i, v := range m {
		ans[i] = v
	}

	return ans
}
